packages/dds/merge-tree/src/segmentPropertiesManager.ts

```diff
@@ -45,133 +47,318 @@ export function copyPropertiesAndManager(
+	const deltas: MapLike<unknown> = {};
+	for (const [key, value] of opToChanges(op, seq)) {
+		run(properties, deltas, key, value);
+		if (properties[key] === null) {
+			// eslint-disable-next-line @typescript-eslint/no-dynamic-delete
+			delete properties[key];
 		}
+	}
+	return deltas;
+}
 
-		const shouldModifyKey = (key: string): boolean => {
-			if (
-				seq === UnassignedSequenceNumber ||
-				seq === UniversalSequenceNumber ||
-				this.pendingKeyUpdateCount?.[key] === undefined
-			) {
-				return true;
-			}
-			return false;
-		};
+/**
+ * The PropertiesManager class handles changes to properties, both remote and local.
+ * It manages the lifecycle for local property changes, ensures all property changes are eventually consistent,
+ * and provides methods to acknowledge changes, update the minimum sequence number (msn), and copy properties to another manager.
+ * This class is essential for maintaining the integrity and consistency of property changes in collaborative environments.
+ * @internal
+ */
+export class PropertiesManager {
+	private readonly changes = new Map<string, PropertyChanges>();
 
-		const deltas: PropertySet = {};
+	/**
+	 * Rolls back local property changes.
+	 * This method reverts property changes based on the provided operation and segment.
+	 * If the operation is part of a collaborative session, it ensures that the changes are consistent with the remote state.
+	 *
+	 * @param op - The operation containing property changes. This can be an adjustment or a set of properties.
+	 * @param seg - The segment containing properties. This object may have a properties map that will be modified.
+	 * @param collaborating - Indicates if the operation is part of a collaborative session. Defaults to false.
+	 * @returns The deltas of the rolled-back properties. This is a map-like object representing the changes that were reverted.
+	 */
+	public rollbackProperties(
+		op: PropsOrAdjust,
+		seg: { properties?: MapLike<unknown> },
+		collaborating: boolean = false,
+	): MapLike<unknown> {
+		return applyChanges(op, seg, UniversalSequenceNumber, (properties, deltas, key, value) => {
+			// eslint-disable-next-line unicorn/no-null
+			const previousValue = properties[key] ?? null;
 
-		for (const [key, newValue] of Object.entries(newProps)) {
-			if (newValue === undefined) {
-				continue;
+			const pending = this.changes.get(key);
+			if (collaborating) {
+				assert(
+					pending !== undefined,
+					"Pending changes must exist for rollback when collaborating",
+				);
+				pending.local.pop();
+				properties[key] = computePropertyValue(
+					pending.msnConsensus,
+					pending.remote.map((n) => n.data),
+					pending.local.map((n) => n.data),
+				);
+			} else {
+				assert(pending === undefined, "Pending changes must not exist when not collaborating");
+				properties[key] = computePropertyValue(previousValue, [value]);
 			}
+			deltas[key] = previousValue;
```

Comment:
since `updateMsn` is the only place where we clean up leftover change entries for keys that no longer have changes, I think we can end up in a strange state if a change is applied locally on a key without an entry in `this.changes`, then roll it back.

Reply:
i mainly exposed this for testing. is the strange case where we have a change entry, but both local and remote are empty? i'll add code to remove the change entry in that case.

Reply:
yeah, that's what I meant. LGTM now.